store/searchtest/user_layer.go

```diff
@@ -6,674 +6,891 @@ package searchtest
-			team.Id,
-			"Tim",
-			&model.UserSearchOptions{
-				AllowFullNames: false,
-				AllowEmails:    true,
-				Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
-			},
-			[]*model.User{},
-		},
-		{
-			"search Bill, search all fields",
-			team.Id,
-			"Bill",
-			&model.UserSearchOptions{
-				AllowFullNames: true,
-				AllowEmails:    true,
-				Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
-			},
-			[]*model.User{u1},
-		},
-		{
-			"search Rob, search all fields",
-			team.Id,
-			"Rob",
-			&model.UserSearchOptions{
-				AllowFullNames: true,
-				AllowEmails:    true,
-				Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
-			},
-			[]*model.User{u1},
-		},
-		{
-			"leading @ should be ignored",
-			team.Id,
-			"@jimb",
-			&model.UserSearchOptions{
-				AllowFullNames: true,
-				Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
-			},
-			[]*model.User{u1},
-		},
-		{
-			"search jim-bobby with system_user roles",
-			team.Id,
-			"jim-bobby",
-			&model.UserSearchOptions{
-				AllowFullNames: true,
-				Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
-				Role:           "system_user",
-			},
-			[]*model.User{u2},
-		},
-		{
-			"search jim with system_admin roles",
-			team.Id,
-			"jim",
-			&model.UserSearchOptions{
-				AllowFullNames: true,
-				Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
-				Role:           "system_admin",
-			},
-			[]*model.User{u1},
-		},
-		{
-			"search ji with system_user roles",
-			team.Id,
-			"ji",
-			&model.UserSearchOptions{
-				AllowFullNames: true,
-				Limit:          model.USER_SEARCH_DEFAULT_LIMIT,
-				Role:           "system_user",
-			},
-			[]*model.User{u1, u2},
-		},
+	options := &model.UserSearchOptions{
+		Limit: model.USER_SEARCH_DEFAULT_LIMIT,
 	}
-
-	for _, testCase := range testCases {
-		testCase := testCase
-		t.Run(testCase.Description, func(t *testing.T) {
-			users, err := s.User().Search(testCase.TeamId, testCase.Term, testCase.Options)
-			require.Nil(t, err)
-			assertUsersMatchInAnyOrder(t, testCase.Expected, users)
-		})
+	t.Run("Should return results when searching for the whole username with underscore", func(t *testing.T) {
+		users, apperr := th.Store.User().AutocompleteUsersInChannel(th.Team.Id, th.ChannelBasic.Id, "alternate_username", options)
+		require.Nil(t, apperr)
+		userAlternate.Sanitize(map[string]bool{})
+		th.assertUsersMatchInAnyOrder(t, []*model.User{userAlternate}, users.InChannel)
+		th.assertUsersMatchInAnyOrder(t, []*model.User{}, users.OutOfChannel)
+	})
+	t.Run("Should return results when searching for part of the username including the underscore", func(t *testing.T) {
+		users, apperr := th.Store.User().AutocompleteUsersInChannel(th.Team.Id, th.ChannelBasic.Id, "_username", options)
+		require.Nil(t, apperr)
+		userAlternate.Sanitize(map[string]bool{})
+		th.assertUsersMatchInAnyOrder(t, []*model.User{userAlternate}, users.InChannel)
+		th.assertUsersMatchInAnyOrder(t, []*model.User{}, users.OutOfChannel)
+	})
+}
```

Comment:
We should add a test case for searching for part of the username not including the underscore, AKA `username` should return `alternate_username`.

The same should happen with `alternate-username` and `alternate.username`

Reply:
Added